### PR TITLE
Make it so all our diagnostics APIs return an independent set of diagnostics.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
In order to get all diagnostics, you must call all the APIs.  And no APIs return
diagnostics produced by other APIs.  This is how things were before the addition
of the getCompletionOptionsDiagnostics API, and i'm returning things to that state.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -105,7 +105,7 @@ namespace ts {
     }
 
     export function getPreEmitDiagnostics(program: Program, sourceFile?: SourceFile): Diagnostic[] {
-        let diagnostics = program.getSyntacticDiagnostics(sourceFile).concat(program.getGlobalDiagnostics()).concat(program.getSemanticDiagnostics(sourceFile));
+        let diagnostics = program.getOptionsDiagnostics().concat(program.getSyntacticDiagnostics(sourceFile)).concat(program.getGlobalDiagnostics()).concat(program.getSemanticDiagnostics(sourceFile));
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Split this into multiple lines

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Also can't concat just take multiple arguments?

File: src/compiler/types.ts [hunk 2]
```diff
@@ -1212,11 +1212,11 @@ namespace ts {
          */
         emit(targetSourceFile?: SourceFile, writeFile?: WriteFileCallback): EmitResult;
 
-        getSyntacticDiagnostics(sourceFile?: SourceFile): Diagnostic[];
+        getOptionsDiagnostics(): Diagnostic[];
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: `getOptionsDiagnostics` sounds vague enough that it's not obvious that it refers to compiler options. Can we not just keep that name?

File: src/services/services.ts [hunk 3]
```diff
@@ -973,6 +973,9 @@ namespace ts {
 
         getSyntacticDiagnostics(fileName: string): Diagnostic[];
         getSemanticDiagnostics(fileName: string): Diagnostic[];
+
+        // TODO: Rename this to getProgramDiagnostics to better indicate that these are any
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: How long do we foresee this TODO sticking around?

File: src/compiler/program.ts (hunk 1 above)
Comment (yuit) [inline src/compiler/program.ts]: I agree with @DanielRosenwasser. Just check ```concat``` take multiple arguments which will clean up the expression a bit

File: src/compiler/types.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: Isn't everything in this API implicitly about the compiler anyways?   Like, we don't have "getCompilerSemanticDiagnostics" :)

File: src/services/services.ts (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/services/services.ts]: Until i can make the change on the managed side and deprecate this entrypoint :)

File: src/compiler/types.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Sure; but we don't have a semantic core to the parser, whereas we do have options that we give when parsing.

File: src/compiler/program.ts (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: Nice!

**Pull request merged**